src/View/View.php

```diff
@@ -687,8 +687,20 @@ public function cache(callable $block, array $options = []): string
         if ($result) {
             return $result;
         }
+
+        $bufferLevel = \ob_get_level();
         ob_start();
-        $block();
+
+        try {
+            $block();
+        } catch (\Throwable $exception) {
+            while (ob_get_level() > $bufferLevel) {
+                ob_end_clean();
```

Comment:
What does this do to error pages that are triggered during view rendering? It looks like you'd end up with a half rendered view and then an error page after that.

Reply:
Good question. From a quick manual test it looks fine.

Is your concern that `ob_get_clean()` could cause output to be generated (it won't), or that discarding the buffers could cause the partially rendered data to be lost?

Reply:
I menr `ob_end_clean()` of course, sorry.

Reply:
Ah I was conflating ob_get_clean() and ob_get_flush() which does flush the output buffer. Discarding the buffer on an exception will be safe.

Reply:
Not that we need to change this, but wouldn't this work as a `finally {}` block?

Reply:
The finally block is run regardless of if an exception is thrown or not. I believe having that run when no exception is thrown would destroy the content output from the view.

Reply:
It would work as in that it would close all opened buffers that way, but it would then always do that, ie also when no exception is being triggered. I didn't go that route because I thought that would mean that it's possibly fixing up and somewhat "hiding" user errors.

If it could be safely assumed that open buffers without exceptions are always user errors, then maybe always closing them and triggering an error, warning, or notice when no exception happened might be a good fit?

Reply:
> If it could be safely assumed that open buffers without exceptions are always user errors, then maybe always closing them and triggering an error, warning, or notice when no exception happened might be a good fit?

I think phpunit does that for us. This seems like it works best for now.

Reply:
It does, though we would have the opportunity to supply more information as to where the problem is located, and what it's possibly caused by. PHPUnit gives you no clue, and it only closes the buffers after the test method ran, so if you for example add debug output in your test method, it won't show up, which can be pretty annoying if you don't know what exactly is going on.

However I just noticed that if you for example leave a view block open (and therefore a buffer), `View::_render()` will trigger an exception telling you about that, and I certainly wouldn't want to interfere with that, so I guess I agree that it's probably best to leave it at that for now 🙃